### PR TITLE
qdisk: preallocate qdisk file after creating header

### PR DESCRIPTION
**Pull request by alltilla**
Related comment: https://github.com/syslog-ng/syslog-ng/pull/4335#discussion_r1135321974

No news file needed.

Signed-off-by: Attila Szakacs <attila.szakacs@axoflow.com>

### CODE REVIEW

File: modules/diskq/qdisk.c [hunk 1]
```diff
@@ -320,12 +320,7 @@ _compat_preallocate(QDisk *self, gint64 size)
   gint64 buf_write_iterations = size / buf_size;
   gint64 additional_write_size = size - buf_write_iterations * buf_size;
 
-  gchar buf[buf_size];
-  for (gint i = 0; i < buf_size; i++)
-    {
-      buf[i] = '\0';
-    }
-  off_t pos = 0;
```
Comment (MrAnno) [inline modules/diskq/qdisk.c:328]: `pos` has been accidentally removed from here.

File: modules/diskq/qdisk.c [hunk 2]
```diff
@@ -343,8 +341,7 @@ _compat_preallocate(QDisk *self, gint64 size)
       return FALSE;
     }
 
-  self->cached_file_size = pos + additional_write_size;
-  g_assert(self->cached_file_size == size);
+  g_assert(pos + additional_write_size == len + offset);
```
Comment (MrAnno) [inline modules/diskq/qdisk.c:318]: `len + offset` are undeclared in this commit.

File: modules/diskq/qdisk.c [hunk 3]
```diff
@@ -296,22 +296,23 @@ _maybe_truncate_file(QDisk *self, gint64 expected_size)
 
 #if !SYSLOG_NG_HAVE_POSIX_FALLOCATE
 static gint
-_compat_preallocate(QDisk *self, gint64 size)
+_compat_preallocate(int fd, off_t offset, off_t len)
 {
   const size_t buf_size = QDISK_RESERVED_SPACE;
-  gint64 buf_write_iterations = size / buf_size;
-  gint64 additional_write_size = size - buf_write_iterations * buf_size;
+  gint64 buf_write_iterations = (len - offset) / buf_size;
```
Comment (MrAnno) [inline modules/diskq/qdisk.c:302]: `len` is not necessarily greater than `offset`.

I think this calculation is incorrect.

> The function posix_fallocate() ensures that disk space is
       allocated for the file referred to by the file descriptor fd for
       the bytes in the range starting at offset and continuing for len
       bytes.

Based on this, we should write exactly `len` bytes after `offset`.

**Comment by MrAnno** [comment]
I've just found this (a bit unrelated, sorry):

`$ man 3p mmap`
> If  the size of the mapped file changes after the call to mmap() as a result of some other operation on the mapped file, the effect of references to portions of the mapped region that correspond to added or removed portions of the file is unspecified.

I think we should reorder a few lines in `_create_header()`.

File: modules/diskq/qdisk.c (hunk 3 above)
Comment (alltilla) [reply modules/diskq/qdisk.c]: Thanks, fixed.

File: modules/diskq/qdisk.c (hunk 2 above)
Comment (alltilla) [reply modules/diskq/qdisk.c]: Yes, I messed up the commit order. Fixed. :)

File: modules/diskq/qdisk.c (hunk 1 above)
Comment (alltilla) [reply modules/diskq/qdisk.c:328]: Thanks, fixed.

**Comment by alltilla** [comment]
> I've just found this (a bit unrelated, sorry):
> 
> `$ man 3p mmap`
> 
> > If  the size of the mapped file changes after the call to mmap() as a result of some other operation on the mapped file, the effect of references to portions of the mapped region that correspond to added or removed portions of the file is unspecified.
> 
> I think we should reorder a few lines in `_create_header()`.

Nice catch, added a new commit, which implements this.

**Review by MrAnno — APPROVED** [review]

**Comment by kira-syslogng** [comment]
Kira-stress-test: Build SUCCESS

**Pull request merged**